sympy/solvers/ode/single.py

```diff
@@ -2695,7 +2700,381 @@ def _get_general_solution(self, *, simplify_flag: bool = True):
+        if coeff == 0:
+            return True
+        if order == 0:
+            if x in coeff.free_symbols:
+                return False
+            return True
+        if coeff.is_Mul:
+            if coeff.has(f(x)):
+                return False
+            return x**order in coeff.args
+        elif coeff.is_Pow:
+            return coeff.as_base_exp() == (x, order)
+        elif order == 1:
+            return x == coeff
+        return False
+
+    def _get_sols(self,r):
+        x = self.ode_problem.sym
+        f = self.ode_problem.func.func
+
+        # First, set up characteristic equation.
+        chareq, symbol = S.Zero, Dummy('x')
+
+        for i in r.keys():
+            if not isinstance(i, str) and i >= 0:
+                chareq += (r[i]*diff(x**symbol, x, i)*x**-symbol).expand()
+
+        chareq = Poly(chareq, symbol)
+        chareqroots = [rootof(chareq, k) for k in range(chareq.degree())]
+        collectterms = []
+
+        # A generator of constants
+        constants = self.ode_problem.get_numbered_constants(num=chareq.degree()*2)
+        constants.reverse()
+
+        # Create a dict root: multiplicity or charroots
+        charroots = defaultdict(int)
+        for root in chareqroots:
+            charroots[root] += 1
+        self.gsol = S.Zero
+        ln = log
+        for root, multiplicity in charroots.items():
+            for i in range(multiplicity):
+                if isinstance(root, RootOf):
+                    self.gsol += (x**root) * constants.pop()
+                    if multiplicity != 1:
+                        raise ValueError("Value should be 1")
+                    collectterms = [(0, root, 0)] + collectterms
+                elif root.is_real:
+                    self.gsol += ln(x)**i*(x**root) * constants.pop()
+                    collectterms = [(i, root, 0)] + collectterms
+                else:
+                    reroot = re(root)
+                    imroot = im(root)
+                    self.gsol += ln(x)**i * (x**reroot) * (
+                        constants.pop() * sin(abs(imroot)*ln(x))
+                        + constants.pop() * cos(imroot*ln(x)))
+                    collectterms = [(i, reroot, imroot)] + collectterms
+
+        self.gsol = Eq(f(x), self.gsol)
+
+        gensols = []
+        # Keep track of when to use sin or cos for nonzero imroot
+        for i, reroot, imroot in collectterms:
+            if imroot == 0:
+                gensols.append(ln(x)**i*x**reroot)
+            else:
+                sin_form = ln(x)**i*x**reroot*sin(abs(imroot)*ln(x))
+                if sin_form in gensols:
+                    cos_form = ln(x)**i*x**reroot*cos(imroot*ln(x))
+                    gensols.append(cos_form)
+                else:
+                    gensols.append(sin_form)
+        return gensols
+
+
+    def _get_general_solution(self, *, simplify_flag: bool = True):
+        self._get_sols(self.r)
+        return [self.gsol]
```

Comment:
Shouldn't this just be `return self._get_sols()`?

It would be better if the methods in these classes don't use internal variables like this. We should minimise instance attributes.

Why is `_get_sols` a different method from `_get_general_solution`?

Reply:
`_get_sols() ` is the method which calcutates the roots of characteristic equation. And to form that we need match_object.
`self.r`  was kept as an argument because this method is also used by euler variation of parameters and euler undetermined coeffs so we need to pass their match object to this.

Reply:
Rather than passing the match object it would be better if this had names arguments like `roots, trialset` or something.

Reply:
> Why can't the vop and undet classes just use the euler class `get_general_solution` method?


because `get_general_solution` will return a equation where as in vop and undet we just want the list of roots of characterstic equation.